qprofiler2/src/org/qcmg/qprofiler2/util/XmlUtils.java

```diff
@@ -0,0 +1,221 @@
+	 * @return
+	 */       
+    public static Element createMetricsNode(Element parent,  String name, Number totalcount ) {
+    	
+    	Element ele = QprofilerXmlUtils.createSubElement( parent,  XmlUtils.metricsEle );
+					
+		if( totalcount != null ) ele.setAttribute( Scount, String.valueOf(totalcount));  
+		 
+		if(name != null) ele.setAttribute( Sname, name );
+		
+		return ele;        	
+    }      
+        
+	/**
+	 * <category name="name" >... </category>
+	 * @param parent
+	 * @param name: category name
+	 * @param id: readgroup id. set to null if not exists
+	 * @return
+	 */
+    public static Element createGroupNode(Element parent, String name) {
+    	Element ele = QprofilerXmlUtils.createSubElement( parent,  XmlUtils.variableGroupEle );	
+    	ele.setAttribute( Sname, name); 
+    	return ele;
+    }
+        
+    /**
+     * output <value name="name">value</value>
+     * @param parent
+     * @param name
+     * @param value
+     */
+    public static <T> Element outputValueNode(Element parent, String name, Number value) {        	
+    	Element ele = QprofilerXmlUtils.createSubElement(parent, Svalue);
+    	ele.setAttribute(Sname, name);
+    	String v = String.valueOf(value);
+
+    	if( value instanceof Double ) {
+    		v = String.format("%,.2f", (double)value);
+    	}
+    	
+    	ele.setTextContent( v );    
+    	return ele;
+    }
+    /**
+     * utput <value name="name" note="comment">value</value>
+     * @param parent
+     * @param name
+     * @param value
+     * @param comment
+     */
+    public static <T> void outputValueNode(Element parent, String name, Number value, String comment) {   
+    	parent.insertBefore( parent.getOwnerDocument().createComment( comment ), parent.getFirstChild() ); 
+    	outputValueNode( parent,  name,  value);
+    	
+    	//parent.appendChild( parent.getOwnerDocument().createComment( name + ": " + comment ) );
+ 	//	Text element = parent.getOwnerDocument().createCDATASection(comment);
+	//	parent.appendChild(element);
+    	    	 
+    }
+            
+    public static void outputBinNode(Element parent, Number start, Number end, Number count ) {
+    	Element ele = QprofilerXmlUtils.createSubElement(parent, Sbin);
+    	ele.setAttribute(Sstart,  String.valueOf(start));
+    	ele.setAttribute(Send,  String.valueOf(end));
+    	ele.setAttribute(Scount, String.valueOf(count)); 
+    }
+   
+    public static <T> void outputTallyGroup( Element parent, String name, Map<T, AtomicLong> tallys, boolean hasPercent ) {
+    	if(tallys == null || tallys.isEmpty()) return;
+    	       	
+    	Element ele = createGroupNode( parent, name);	//<category>       	
+		double sum = hasPercent ? tallys.values().stream().mapToDouble( x -> (double) x.get() ).sum() : 0;	
+		
+		for(Entry<T,  AtomicLong> entry : tallys.entrySet()) { 
+			//skip zero value for output
+			if(entry.getValue().get() == 0 ) continue;
+			Double percent = (sum == 0)? null : 100 * (double)entry.getValue().get() / sum;
```

Comment:
can you use the primitive rather than the object?

Reply:
sure, corrected